drivers/sensor/ms5611/ms5611.c

```diff
@@ -0,0 +1,551 @@
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	int ret;
+
+	if (attr != SENSOR_ATTR_OVERSAMPLING) {
+		return -ENOTSUP;
+	}
+
+	switch (chan) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		ret = ms5611_temperature_osr_set(dev, val);
+		break;
+	case SENSOR_CHAN_PRESS:
+		ret = ms5611_pressure_osr_set(dev, val);
+		break;
+	case SENSOR_CHAN_ALL:
+		ret = ms5611_pressure_osr_set(dev, val);
+		if (ret != 0)
+			break;
+
+		ret = ms5611_temperature_osr_set(dev, val);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+	return ret;
+}
+
+static int ms5611_fetch_temp_and_press(const struct device *dev)
+{
+	const struct ms5611_config *cfg = dev->config;
+	struct ms5611_data *data = dev->data;
+	uint8_t temp_data[3], press_data[3];
+	uint32_t raw_temp, raw_press;
+	int ret;
+	int32_t dT, temp, t2, p;
+	uint8_t i2c_cmd;
+	int64_t off, off2, sens, sens2;
+
+	/* Request temperature conversion */
+	i2c_cmd = data->osr[MS5611_OSR_TEMP_IDX].read_cmd;
+	ret = i2c_write_dt(&cfg->i2c_bus, &i2c_cmd, sizeof(i2c_cmd));
+	if (ret < 0) {
+		LOG_ERR("Failed to request temperature conversion");
+		return ret;
+	}
+
+	/* Wait for conversion to be finished */
+	k_sleep(K_USEC(data->osr[MS5611_OSR_TEMP_IDX].resp_time));
+
+	i2c_cmd = MS5611_CMD_ADC_READ;
+	ret = i2c_write_dt(&cfg->i2c_bus, &i2c_cmd, sizeof(i2c_cmd));
+	if (ret < 0) {
+		LOG_ERR("Failed to request temperature read");
+		return ret;
+	}
+
+	ret = i2c_read_dt(&cfg->i2c_bus, temp_data, sizeof(temp_data));
+	if (ret < 0) {
+		LOG_ERR("Failed to read temperature measurement");
+		return ret;
+	}
+
+	raw_temp = sys_get_be24(temp_data);
+
+	if (raw_temp == 0) {
+		LOG_DBG("Invalid temperature data obtained");
+		return -EIO;
+	}
+
+	/* Request pressure conversion */
+	i2c_cmd = data->osr[MS5611_OSR_PRES_IDX].read_cmd;
+	ret = i2c_write_dt(&cfg->i2c_bus, &i2c_cmd, sizeof(i2c_cmd));
+	if (ret < 0) {
+		LOG_ERR("Failed to request pressure conversion");
+		return ret;
+	}
+
+	/* Wait for conversion to be finished */
+	k_sleep(K_USEC(data->osr[MS5611_OSR_PRES_IDX].resp_time));
+
+	i2c_cmd = MS5611_CMD_ADC_READ;
+	ret = i2c_write_dt(&cfg->i2c_bus, &i2c_cmd, sizeof(i2c_cmd));
+	if (ret < 0) {
+		LOG_ERR("Failed to request pressure read");
+		return ret;
+	}
+
+	ret = i2c_read_dt(&cfg->i2c_bus, press_data, sizeof(press_data));
+	if (ret < 0) {
+		LOG_ERR("Failed to read pressure measurement");
+		return ret;
+	}
+
+	raw_press = sys_get_be24(press_data);
+
+	if (raw_press == 0) {
+		LOG_DBG("Invalid pressure data obtained");
+		return -EIO;
+	}
+
+	/* Calculate compensated temperature value */
+	dT = raw_temp - (data->prom[5] << 8);
+	temp = 2000 + (((int64_t)dT * data->prom[6]) >> 23);
+
+	/* Second order temperature compensation */
+	if (temp < 2000) {
+		t2 = ((dT ^ 2) >> 31);
+		off2 = 5 * ((temp - 2000) ^ 2) / 2;
+		sens2 = 5 * (temp - 2000) ^ 2 / 4;
+
+		if (temp < -1500) {
+			off2 = off2 + (7 * (temp + 1500) ^ 2);
+			sens2 = sens2 + (11 * ((temp + 1500) ^ 2)) / 2;
+		}
```

Comment:
Please document the source of this temperature compensation algorithm. Link to a datasheet and section number, for example.

Reply:
Added few lines above